core/kernel/src/c_interface.rs

```diff
@@ -0,0 +1,129 @@
+use crate::{
+    feature_manifest::{FeatureManifest, FeatureManifestParsingError},
+    kernel::Kernel,
+    kernel::KERNEL,
+};
+use serde::Serialize;
+use std::{
+    ffi::{CStr, CString},
+    os::raw::c_char,
+    str::Utf8Error,
+};
+
+/// # Safety
+/// This function can be unsafe if the language_code pointer is null or the string is in wrong format.
+///
+/// Bootstrap the kernel with the given configuration:
+/// - language_code: User's locale language code
+/// Returns the result JSON of either success or failure to bootstrap the kernel
+#[no_mangle]
+pub unsafe extern "C" fn kernel_bootstrap(language_code: *const c_char) -> *const c_char {
+    unsafe fn bootstrap(language_code: *const c_char) -> Result<(), KernelError> {
+        let language_code = cstring_to_str(&language_code).map(String::from)?;
+        Kernel::bootstrap(language_code).map_err(KernelError::FailedToBootstrapKernel)
+    }
+
+    to_result_str(bootstrap(language_code))
+}
+
+/// # Safety
+/// This function can be unsafe if the json pointer is null or the string is in wrong format.
+///
+/// Parse the feature manifest from a given JSON.
+/// Returns the JSON with the parsed feature manifest by taking into account the user's locale language code.
+///
+/// This is unfortunate at the moment that this API gets a JSON and responds back with a JSON. Ideally the kernel
+/// would be able to read all the features from the disk on its own and return back a JSON with all of the features.
+/// It was choosen to return back to platform code a JSON, as composing complex C model is not nice. For example
+/// Rust's HashMap cannot be directly mapped to C, and instead of working with Opaque types and pointers to access data from a hashmap - use the JSON.
+/// The JSON would allow to easily represent any complex data model in a standard way.
+///
+/// This API would be an example of any further API that will be added. That is - the API would respond back with a JSON
```

Comment:
What about using a protobuf? There's a C implementation.
The problem I see with returning a JSON is that we're not really earning much with the platform code. That JSON will have to be interpreted by the app code again, with all due checks and so on. It's not native in any way, so the only functionality we're getting out of this is a pre-interpretation (that can be skipped, since we will actually need a manifest parser on the receiving end anyway)

Reply:
See my comment above. I discovered FlatBuffers after that. Still, point is the same: return a format that needs not be reinterpreted on the other end.

Reply:
Sure the Protobuf can be an option. But, in my understanding, sending it over C bridge would anyway require serialisation, if not to a string like JSON than to binary data. To me the only thing we get with Protobuf is that we have the shared schema implementation.

Be it with C structs, JSON or Protobufs we have to have some sort of re-interpretation in the Swift/Kotlin code. With pure C, it's (possibly) less interpretation, but still we would need some. For example a `const char *` has to be mapped to a Swift String before usage. Same is true for an array/vector. We would have to map unsafe pointers to safe data structures. I don't see it feasible to use directly the C data structure throughout the Kotlin/Swift code.

Reply:
What about FlatBuffers? My understanding is that it will help preserve data structures natively. Point is, JSON to JSON is not a great trip. If we were going to use those data structures natively, it would be OK; for instance, we could think about a higher-level access to the manifest that would totally encapsulate data structures and that would not need to expose all those implementation details. But if we positively need to, and this is not limited to manifest, we would need a better way to represent any kind of data structures that travels well over the C bridge. Although I haven't had time to investigate it closely, FlatBuffer seems a good option, covering all languages we need (also Kotlin)

Reply:
I will have to look and see the differences between Protobufs and Flatbuffers. But as I've said the main benefit of those is that we have one schema defined, then 4 variants are generated from it for Swift/Kotlin/JS/Rust. This can help with making sure that are models are in sync.

I will experiment with one those two as opposed to JSON. But the main idea in the end is that we would need some kind of serialisation to avoid fighting with C limitations.